system/lib/wasmfs/syscalls.cpp

```diff
@@ -1104,4 +1112,66 @@ long __syscall_ftruncate64(long fd, long low, long high) {
   }
   return ret;
 }
+
+long __syscall_pipe(long fd) {
+  auto* fds = (__wasi_fd_t*)fd;
+
+  // A global singleton PipeBackend is used for all pipes.
+  static auto backend = wasmFS.addBackend(std::make_unique<PipeBackend>());
+
+  // Make a pipe: Two PipeFiles that share a single data source between them, so
+  // that writing to one can be read in the other.
+  auto data = std::make_shared<PipeData>();
+  auto reader = std::make_shared<PipeFile>(S_IRUGO, backend, data);
+  auto writer = std::make_shared<PipeFile>(S_IWUGO, backend, data);
+
+  auto fileTable = wasmFS.getFileTable().locked();
+  fds[0] =
+    fileTable.addEntry(std::make_shared<OpenFileState>(0, O_RDONLY, reader));
+  fds[1] =
+    fileTable.addEntry(std::make_shared<OpenFileState>(0, O_WRONLY, writer));
+
+  return 0;
+}
+
+int __syscall_poll(struct pollfd* fds, nfds_t nfds, int timeout) {
```

Comment:
Can we separate this into a follow-up?

Reply:
I'd rather not as it would require more surgery in the test to disable more parts and then to re-enable them later.

My general position is that we have so much stuff to get done, and that bisectability does not matter yet (not until a lot more works), so it's ok to bundle PRs.

But I don't feel strongly, and I can see the logic of separating it out, of course.